Make CONJECTURE_ASSERT crash in WTFCrashDueToConjectureAssert.

MenloDorian opened the pull request:
#### 5ed2b87de57716b0af4ffc7f36143a3fa73a4733
<pre>
Make CONJECTURE_ASSERT crash in WTFCrashDueToConjectureAssert.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279562">https://bugs.webkit.org/show_bug.cgi?id=279562</a>
<a href="https://rdar.apple.com/135839946">rdar://135839946</a>

Reviewed by Alexey Proskuryakov.

This will make it easier to distinguish from other normal assertion failures.

1. Also changed the failure message and dumped the stack.  For reference, here&apos;s an
   example of a failure message which I fabricated by inserting a bad CONJECTURE_ASSERT:

    CONJECTURE ASSERTION FAILED: vmCreationShouldCrash
    ./runtime/VM.cpp(244) : JSC::VM::VM(VMType, HeapType, WTF::RunLoop *, bool *)
    1   0x103389e34 WTFCrashDueToConjectureAssert
    2   0x1041f869c JSC::VM::VM(JSC::VM::VMType, JSC::HeapType, WTF::RunLoop*, bool*)
    3   0x1041f9768 JSC::VM::tryCreate(JSC::HeapType, WTF::RunLoop*)
    4   0x100df5690 jscmain(int, char**)
    5   0x100df519c main
    6   0x19ea38274 start

2. Also added a runtime switch for enabling / disabling CONJECTURE_ASSERTs.  Note: there
   is already a build time flag ENABLE_CONJECTURE_ASSERT that needs to be set to 1 in order
   for CONJECTURE_ASSERTs to be supported.  In addition to that, CONJECTURE_ASSERTs is
   disabled by default at runtime.  To enable it, you need to define the environment
   variable ENABLE_WEBKIT_CONJECTURE_ASSERT.

A reminder about using CONJECTURE_ASSERT: it is not enable in the EWS builds, and hence,
is not tested by the EWS.  The person who adds a CONJECTURE_ASSERT is responsible for testing
that it builds (with ENABLE_CONJECTURE_ASSERT=1) and that it runs (passes tests).

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):

Canonical link: <a href="https://commits.webkit.org/283648@main">https://commits.webkit.org/283648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4edcfe9cfc7aff7e0d1d63de23d0e38b26bfb14

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66939 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/46314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/18072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/54113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/17853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/70974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/59/builds/18072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70006 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/54113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/70974 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/54113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/16426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/60053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/54113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66184 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/10896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/17853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/72675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/10928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/72675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/19561 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/87952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10151 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42121 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/87952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/43198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/44381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/42941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WTF/wtf/Threading.cpp:
```diff
@@ -500,6 +500,9 @@ void initialize()
     static std::once_flag onceKey;
     std::call_once(onceKey, [] {
+#if ENABLE(CONJECTURE_ASSERT)
+        wtfConjectureAssertIsEnabled = !!getenv("ENABLE_WEBKIT_CONJECTURE_ASSERT");
```
Do environment variables get propagated to child processes of all kinds, or do we need to propagate them from UI process ourselves?

ddkilzer replied:
I don't think we should be in the business of propagating environment variables.  The user needs to know how to set `__XPC_` environment variables if they want to target XPC processes.

MenloDorian replied:
I've confirmed that `__XPC_` will do the job of passing the env var on.

aproskuryakov commented on Source/WTF/wtf/Assertions.cpp:
```diff
@@ -374,6 +374,18 @@ void WTFCrash()
 #endif // !defined(NDEBUG) || !(OS(DARWIN) || PLATFORM(PLAYSTATION))
 
+#if ENABLE(CONJECTURE_ASSERT)
+int wtfConjectureAssertIsEnabled = 0;
```
Is there a reason why it's not a `bool`?

MenloDorian replied:
This variable needs to be declared in Assertions.h, and Assertions.h need to be able to work with C code as well.

aproskuryakov replied:
We already use `bool` in this file, and include `stdbool.h`.

```
WTF_EXPORT_PRIVATE bool WTFWillLogWithLevel(WTFLogChannel*, WTFLogLevel);
```

aproskuryakov commented on Source/WTF/wtf/Assertions.h:
```diff
@@ -765,9 +770,19 @@ constexpr bool assertionFailureDueToUnreachableCode = false;
    stay in the codebase perpetually.
- */
+
+   Note: there is no EWS coverage for CONJECTURE_ASSERTs. So, if you add one, you are
```
Please consider removing the "note" prefix. Every comment is a note.

I also have a nitpick for the comment above - "If the conjecture is proven false it" seems to have some words left in accidentally while editing.

MenloDorian replied:
I normally use "Note:" to call extra attention to specific points.  However, in this case, after re-reading the comment in entirety, there's no reason that this one detail should stand out over the others.  So, I removed it ... and fix the other typo.

aproskuryakov reviewed (approved):
r=me with comments, assuming that propagation to WebContent and other processes actually works.

Commit:
Make CONJECTURE_ASSERT crash in WTFCrashDueToConjectureAssert.
https://bugs.webkit.org/show_bug.cgi?id=279562
rdar://135839946

Reviewed by Alexey Proskuryakov.

This will make it easier to distinguish from other normal assertion failures.

1. Also changed the failure message and dumped the stack.  For reference, here's an
   example of a failure message which I fabricated by inserting a bad CONJECTURE_ASSERT:

    CONJECTURE ASSERTION FAILED: vmCreationShouldCrash
    ./runtime/VM.cpp(244) : JSC::VM::VM(VMType, HeapType, WTF::RunLoop *, bool *)
    1   0x103389e34 WTFCrashDueToConjectureAssert
    2   0x1041f869c JSC::VM::VM(JSC::VM::VMType, JSC::HeapType, WTF::RunLoop*, bool*)
    3   0x1041f9768 JSC::VM::tryCreate(JSC::HeapType, WTF::RunLoop*)
    4   0x100df5690 jscmain(int, char**)
    5   0x100df519c main
    6   0x19ea38274 start

2. Also added a runtime switch for enabling / disabling CONJECTURE_ASSERTs.  Note: there
   is already a build time flag ENABLE_CONJECTURE_ASSERT that needs to be set to 1 in order
   for CONJECTURE_ASSERTs to be supported.  In addition to that, CONJECTURE_ASSERTs is
   disabled by default at runtime.  To enable it, you need to define the environment
   variable ENABLE_WEBKIT_CONJECTURE_ASSERT.

A reminder about using CONJECTURE_ASSERT: it is not enable in the EWS builds, and hence,
is not tested by the EWS.  The person who adds a CONJECTURE_ASSERT is responsible for testing
that it builds (with ENABLE_CONJECTURE_ASSERT=1) and that it runs (passes tests).

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/Assertions.h:
* Source/WTF/wtf/Threading.cpp:
(WTF::initialize):

Canonical link: https://commits.webkit.org/283648@main
